tools/generate_config.py

```diff
@@ -0,0 +1,326 @@
+#!/usr/bin/env python3
+
+import argparse
+from pathlib import Path
+import json
+try:
+    import jsonschema
+except ImportError:
+    jsonschema = None
+from collections import OrderedDict
+
+THIS_DIR = Path(__file__).parent
+
+if (THIS_DIR / 'schemas').exists():
+    SCHEMAS_DIR = THIS_DIR / 'schemas'
+else:
+    SCHEMAS_DIR = THIS_DIR.parent / 'share' / 'schemas'
+
+ENCLAVE_CFG_SCHEMA_PATH = SCHEMAS_DIR / 'enclave-config.schema.json'
+
+# Validates a particular enclave_cfg_path against ENCLAVE_CFG_SCHEMA_PATH
+def validate(args):
+    if jsonschema is None:
+        raise RuntimeError('validate requires the "jsonschema" Python package')
+    schema_path = ENCLAVE_CFG_SCHEMA_PATH
+    print('Validating {args.file}')
+    with open(args.file) as f:
+      cfg = json.load(f)
+    with open(schema_path) as f:
+      cfg_schema = json.load(f)
+    jsonschema.validate(instance=cfg, schema=cfg_schema)
+    print('No errors found.')
+
+def post_type(jtype):
+  if 'type' in jtype:
+    jtt = jtype['type']
+    if 'maxLength' in jtype:
+      if jtt == 'string':
+        return '[' + str(jtype['maxLength']+1) + ']'
+      elif jtt == 'array':
+        return '[' + str(jtype['maxLength']) + ']'
+    return ''
+  else:
+    return ''
+
+def pre_type(jtype):
+  if 'type' in jtype:
+    jtt = jtype['type']
+    if jtt == 'array' or (isinstance(jtt, list) and 'array' in jtt and 'null' in jtt):
+      items = jtype['items']
+      item_type = pre_type(items)
+      if 'maxLength' in jtype:
+        return item_type
+      else:
+        return item_type + '*'
+    elif jtt == 'boolean': return 'bool'
+    elif jtt == 'none': return 'null'
+    elif jtt == 'string' or (isinstance(jtt, list) and 'string' in jtt and 'null' in jtt):
+      return 'char' if 'maxLength' in jtype else 'char*'
+    else:
+        print('unhandled json type: %s' % jtype)
+  elif '$ref' in jtype:
+    rtype = jtype['$ref'][jtype['$ref'].rfind('/')+1:]
+    if rtype.startswith('safe_'):
+      return rtype[5:]
+    else:
+      return rtype
+  else:
+    print('unknown json type: %s' % jtype)
+    return jtype
+
+def need_size_var(jtype):
+  if 'type' not in jtype:
+    return False
+  jtt = jtype['type']
+  return (jtt == 'array' and 'maxLength' not in jtype) or (isinstance(jtt, list) and 'array' in jtt and 'null' in jtt)
+
+num_settings = 0
+
+def generate_header(schema_file_name, root, args):
+  global num_settings
+
+  with open(str(args.header), "w") as header:
+    header.write('#ifndef SGXLKL_ENCLAVE_CONFIG_GEN_H\n');
```

Comment:
You could remove noise by first writing everything into a string with `s += '..'` and then write it to file in one go.

Reply:
Yeah, there are plenty of ways to improve this hacky script, but I don't feel it's worth the effort. If we ever need some new feature in this, we should really switch to a proper, fully-featured schema code extractor (there are apparently lots of them).

Reply:
Good to know, did you find one for C? In the list at https://json-schema.org/implementations.html#code-generation there's none.

Reply:
Yes, IBM has one: https://www.ibm.com/support/knowledgecenter/SSGMCP_5.2.0/com.ibm.cics.ts.applicationprogramming.doc/datamapping/dfhws_json2c.html

Reply:
Seems this is specifically for their web service bindings, and the software isn't free.

Reply:
Entirely possible; you have read more about it then I have.